core/modules/replica/tools/qserv-replica-master-http.cc

```diff
@@ -0,0 +1,380 @@
+            [self] (std::string const& worker2evict) {
+                self->_evict(worker2evict);
+            },
+            _workerEvictTimeoutSec,
+            _workerResponseTimeoutSec,
+            _healthProbeIntervalSec
+        );
+        _healthMonitorThread->start();
+
+        _httpThread = HttpThread::create(
+            _controller,
+            [self] (ControlThread::Ptr const& ptr) {
+                self->_failed = true;
+            },
+            [self] (std::string const& worker2evict) {
+                self->_evict(worker2evict);
+            },
+            _healthMonitorThread,
+            _replicationThread,
+            _deleteWorkerThread
+        );
+        _httpThread->start();
+
+        // Keep running before a catastrophic failure is reported by any
+        // above initiated activity
+
+        util::BlockPost blockPost(1000, 2000);
+        while (not _failed) {
+            blockPost.wait();
+        }
+
+        // Stop all threads if any are still running
+
+        _healthMonitorThread->stop();
+        _replicationThread->stop();
+
+        return 1;
+    }
+
+private:
+
+    /**
+     * Evict the specified worker from the cluster
+     *
+     * NOTE: This method is called by the health-monitoring thread when
+     * a condition for evicting the worker is detected. The calling thread
+     * will be blocked for a duration of this call.
+     *
+     * @param worker
+     *   the name of a worker to be evicted
+     */
+    void _evict(std::string const& worker) {
+
+        // This thread needs to be stopped to avoid any interference with
+        // the worker exclusion protocol.
+
+        _replicationThread->stop();
+
+        // This thread will be allowed to run for as long as it's permitted by
+        // the corresponding timeouts set for Requests and Jobs in the Configuration,
+        // or until a catastrophic failure occurs within any control thread (including
+        // this one).
+
+        auto self = shared_from_base<MasterControllerHttpApp>();
+
+        _deleteWorkerThread = DeleteWorkerThread::create(
+            _controller,
+            [self] (ControlThread::Ptr const& ptr) {
+                self->_failed = true;
+            },
+            worker,
+            _permanentDelete
+        );
+        _deleteWorkerThread->startAndWait(
+            [self] (ControlThread::Ptr const& ptr) -> bool {
+                return self->_failed;
+            }
+        );
+        _deleteWorkerThread->stop();    // it's safe to call this method even if the thread is
+                                        // no longer running.
+
+        _deleteWorkerThread = nullptr;  // the object is no longer needed because it was
+                                        // created for a specific worker.
+
+        // Resume the normal replication sequence unless a catastrophic failure
+        // in the system has been detected
+        
+        if (not _failed) _replicationThread->start();
```

Comment:
I'm not seeing how it's possible to be confident in a `ControlThread` that `start` can't be called while `stop` is pending.

Reply:
ControlThread's methods *start* and *stop* have a thread-safe implementation. Note the *lock* on a mutex acquired by each method before attempting a state transition. Here is:
```
bool ControlThread::start() {
    util::Lock lock(_mtx, context() + "start");
    if (_isRunning.exchange(true)) {
        return true;
    }
    ..
}


bool ControlThread::stop() {
    util::Lock lock(_mtx, context() + "stop");
    if (not _isRunning) return true;
    _stopRequested = true;
    ....
}
```